[1.11-rc2] multiple panics around json parsing

This is an upgraded environment from 1.10.3.  After running for slight bit I got this panic.

```
unexpected fault address 0x0
fatal error: fault
[signal 0xb code=0x80 addr=0x0 pc=0x6912f9]

goroutine 218 [running]:
runtime.throw(0x17aab90, 0x5)
	/usr/local/go/src/runtime/panic.go:527 +0x90 fp=0xc8212e7718 sp=0xc8212e7700
runtime.sigpanic()
	/usr/local/go/src/runtime/sigpanic_unix.go:27 +0x2ab fp=0xc8212e7768 sp=0xc8212e7718
encoding/json.unquoteBytes(0xc820d77a8d, 0x4b7, 0xe4f, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/encoding/json/decode.go:1006 +0x119 fp=0xc8212e7840 sp=0xc8212e7768
encoding/json.(*decodeState).literalStore(0xc82110da40, 0xc820d77a8c, 0x4b7, 0xe50, 0x12af160, 0xc820fd7838, 0xd8, 0x0)
	/usr/local/go/src/encoding/json/decode.go:738 +0xb4e fp=0xc8212e7ce8 sp=0xc8212e7840
encoding/json.(*decodeState).literal(0xc82110da40, 0x12af160, 0xc820fd7838, 0xd8)
	/usr/local/go/src/encoding/json/decode.go:646 +0xe9 fp=0xc8212e7d50 sp=0xc8212e7ce8
encoding/json.(*decodeState).value(0xc82110da40, 0x12af160, 0xc820fd7838, 0xd8)
	/usr/local/go/src/encoding/json/decode.go:299 +0x397 fp=0xc8212e7e28 sp=0xc8212e7d50
encoding/json.(*decodeState).object(0xc82110da40, 0x1632b80, 0xc820fd7810, 0xd9)
	/usr/local/go/src/encoding/json/decode.go:613 +0x1179 fp=0xc8212e81d0 sp=0xc8212e7e28
encoding/json.(*decodeState).value(0xc82110da40, 0x1632b80, 0xc820fd7810, 0xd9)
	/usr/local/go/src/encoding/json/decode.go:296 +0x3bd fp=0xc8212e82a8 sp=0xc8212e81d0
encoding/json.(*decodeState).array(0xc82110da40, 0x1274000, 0xc820d73858, 0xd7)
	/usr/local/go/src/encoding/json/decode.go:447 +0xa6b fp=0xc8212e84b0 sp=0xc8212e82a8
encoding/json.(*decodeState).value(0xc82110da40, 0x1274000, 0xc820d73858, 0xd7)
	/usr/local/go/src/encoding/json/decode.go:293 +0x3dd fp=0xc8212e8588 sp=0xc8212e84b0
encoding/json.(*decodeState).object(0xc82110da40, 0x1654d00, 0xc820d73680, 0xd9)
	/usr/local/go/src/encoding/json/decode.go:613 +0x1179 fp=0xc8212e8930 sp=0xc8212e8588
encoding/json.(*decodeState).value(0xc82110da40, 0x1646600, 0xc820d73680, 0x16)
	/usr/local/go/src/encoding/json/decode.go:296 +0x3bd fp=0xc8212e8a08 sp=0xc8212e8930
encoding/json.(*decodeState).unmarshal(0xc82110da40, 0x1646600, 0xc820d73680, 0x0, 0x0)
	/usr/local/go/src/encoding/json/decode.go:157 +0x196 fp=0xc8212e8ae0 sp=0xc8212e8a08
encoding/json.Unmarshal(0xc820d77500, 0x11dc, 0x13dc, 0x1646600, 0xc820d73680, 0x0, 0x0)
	/usr/local/go/src/encoding/json/decode.go:85 +0x12b fp=0xc8212e8b50 sp=0xc8212e8ae0
github.com/docker/docker/image.NewFromJSON(0xc820d77500, 0x11dc, 0x13dc, 0xc820d77500, 0x0, 0x0)
	/go/src/github.com/docker/docker/image/image.go:128 +0xaf fp=0xc8212e8be0 sp=0xc8212e8b50
github.com/docker/docker/image.(*store).Get(0xc820013280, 0xc8207b8910, 0x47, 0xc8212e8cb0, 0x0, 0x0)
	/go/src/github.com/docker/docker/image/store.go:189 +0xee fp=0xc8212e8c78 sp=0xc8212e8be0
github.com/docker/docker/image.(*store).imagesMap(0xc820013280, 0xc82178d601, 0x0)
	/go/src/github.com/docker/docker/image/store.go:287 +0x1f6 fp=0xc8212e8d98 sp=0xc8212e8c78
github.com/docker/docker/image.(*store).Map(0xc820013280, 0xc821046b40)
	/go/src/github.com/docker/docker/image/store.go:274 +0x26 fp=0xc8212e8db8 sp=0xc8212e8d98
github.com/docker/docker/daemon.(*Daemon).Images(0xc8206f4a80, 0xc821046a80, 0x2a, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/docker/docker/daemon/images.go:63 +0x142b fp=0xc8212e9208 sp=0xc8212e8db8
github.com/docker/docker/api/server/router/image.(*imageRouter).getImagesJSON(0xc820758fc0, 0x7f7c0f746570, 0xc8210469c0, 0x7f7c0f74c700, 0xc821576370, 0xc82254ed20, 0xc821046900, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/router/image/image_routes.go:313 +0x1a7 fp=0xc8212e9348 sp=0xc8212e9208
github.com/docker/docker/api/server/router/image.(*imageRouter).(github.com/docker/docker/api/server/router/image.getImagesJSON)-fm(0x7f7c0f746570, 0xc8210469c0, 0x7f7c0f74c700, 0xc821576370, 0xc82254ed20, 0xc821046900, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/router/image/image.go:29 +0x74 fp=0xc8212e9398 sp=0xc8212e9348
github.com/docker/docker/api/server/middleware.NewVersionMiddleware.func1.1(0x7f7c0f746570, 0xc8210469c0, 0x7f7c0f74c700, 0xc821576370, 0xc82254ed20, 0xc821046900, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/middleware/version.go:42 +0xa43 fp=0xc8212e9548 sp=0xc8212e9398
github.com/docker/docker/api/server/middleware.NewUserAgentMiddleware.func1.1(0x7f7c0f746570, 0xc821046990, 0x7f7c0f74c700, 0xc821576370, 0xc82254ed20, 0xc821046900, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/middleware/user_agent.go:34 +0x6b8 fp=0xc8212e96d8 sp=0xc8212e9548
github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1(0x7f7c0e6b9068, 0xc82000f030, 0x7f7c0f74c700, 0xc821576370, 0xc82254ed20, 0xc821046900, 0x0, 0x0)
	/go/src/github.com/docker/docker/api/server/middleware/debug.go:21 +0xd57 fp=0xc8212e9970 sp=0xc8212e96d8
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1(0x7f7c0f74c700, 0xc821576370, 0xc82254ed20)
	/go/src/github.com/docker/docker/api/server/server.go:132 +0x138 fp=0xc8212e9a68 sp=0xc8212e9970
net/http.HandlerFunc.ServeHTTP(0xc8209d5b80, 0x7f7c0f74c700, 0xc821576370, 0xc82254ed20)
	/usr/local/go/src/net/http/server.go:1422 +0x3a fp=0xc8212e9a88 sp=0xc8212e9a68
github.com/gorilla/mux.(*Router).ServeHTTP(0xc820a625a0, 0x7f7c0f74c700, 0xc821576370, 0xc82254ed20)
	/go/src/github.com/docker/docker/vendor/src/github.com/gorilla/mux/mux.go:98 +0x29e fp=0xc8212e9ba0 sp=0xc8212e9a88
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP(0xc82095f0e0, 0x7f7c0f74c700, 0xc821576370, 0xc82254ed20)
	/go/src/github.com/docker/docker/api/server/router_swapper.go:29 +0x72 fp=0xc8212e9bd0 sp=0xc8212e9ba0
net/http.serverHandler.ServeHTTP(0xc82001ab40, 0x7f7c0f74c700, 0xc821576370, 0xc82254ed20)
	/usr/local/go/src/net/http/server.go:1862 +0x19e fp=0xc8212e9c30 sp=0xc8212e9bd0
net/http.(*conn).serve(0xc8215762c0)
	/usr/local/go/src/net/http/server.go:1361 +0xbee fp=0xc8212e9f88 sp=0xc8212e9c30
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1721 +0x1 fp=0xc8212e9f90 sp=0xc8212e9f88
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1910 +0x3f6
```

After that panic I can no longer start Docker and get

```
./src/docker/bundles/latest/binary/docker daemon -D -s overlay
DEBU[0000] docker group found. gid: 1101                
DEBU[0000] Listener created for HTTP on unix (/var/run/docker.sock) 
INFO[0000] previous instance of containerd still alive (9622) 
DEBU[0000] containerd connection state change: READY    
DEBU[0000] Using default logging driver json-file       
DEBU[0000] Golang's threads limit set to 111780         
DEBU[0000] [graphdriver] trying provided driver "overlay" 
DEBU[0000] Using graph driver overlay                   
DEBU[0000] received past containerd event: &types.Event{Type:"exit", Id:"16f659ec9d86d640a267dc212dc0581f5f27a5b55c232d3b1e847a627eb0f23e", Status:0x89, Pid:"init", Timestamp:0x56fa9398} 
DEBU[0000] received past containerd event: &types.Event{Type:"start-container", Id:"520ad5f2af7b7032a26f26870d982d3015c603dd06bc414d151e5a8bd9594553", Status:0x0, Pid:"", Timestamp:0x56fa9398} 
DEBU[0000] received past containerd event: &types.Event{Type:"exit", Id:"c1d8f65990a0275fab1ca9623c911f5ede069a37ee85305538db52b8c2615e4b", Status:0x0, Pid:"init", Timestamp:0x56fa9398} 
DEBU[0000] received past containerd event: &types.Event{Type:"live", Id:"", Status:0x0, Pid:"", Timestamp:0x56fa9402} 
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x20]

goroutine 1 [running]:
encoding/json.(*decodeState).unmarshal.func1(0xc820de8878)
	/usr/local/go/src/encoding/json/decode.go:143 +0x88
time.Parse(0x18ebb40, 0x1b, 0xc820cab1a0, 0x20, 0xece30be4b, 0x796a218, 0x2250b80, 0x0, 0x0)
	/usr/local/go/src/time/format.go:709 +0x86
```

```
 docker version
Client:
 Version:      1.11.0-rc2
 API version:  1.23
 Go version:   go1.5.3
 Git commit:   388f544-unsupported
 Built:        Tue Mar 29 14:35:27 2016
 OS/Arch:      linux/amd64
```
(I can't get the server to start to get the server portion, but this is built from source using the 1.11.0-rc2 tag)

--- a/layer/layer_store.go
+++ b/layer/layer_store.go
@@ -334,7 +334,10 @@ func (ls *layerStore) get(l ChainID) *roLayer {
 }
 
 func (ls *layerStore) Get(l ChainID) (Layer, error) {
-	layer := ls.get(l)
+	ls.layerL.Lock()
+	defer ls.layerL.Unlock()
+
+	layer := ls.getWithoutLock(l)
 	if layer == nil {
 		return nil, ErrLayerDoesNotExist
 	}

